### ISSUE TITLE
Libresonic v6.3 Planning

### ISSUE THREAD

**Issue by EugeneKay** [Libresonic/libresonic#373]
This will be the Omnibus Release Planning thread for Libresonic 6.3. This is a developer discussion; please limit user discussions/issues to IRC or other Issues. This will be a development release, meaning that there will be a lot of breakage in the develop branch before it is stable. Questions to be answered:

- What is the release date: Planned for July 4th, 2017(approximately two months from 6.2)
- What features are targeted for this release? Suggestions: removal/replacement of JWPlayer, Windows/Linux packaging
- Are there any outstanding issues from 6.2 that should go into develop instead of stable(to become 6.2.1)?
- [Do you want a :taco:?](https://youtu.be/k3tcQI5EEiA)

Please indicate what you would like to work on for this release, so that we can get an estimate of what will actually happen(in the form of a checklist).

**Comment by biconou** [comment]
According to recent discussions, can I suggest : 

- refactor of the playQueue view with a better code organisation and better javascript practices.
- eventually, use a better technique to update the play queue status (websocket ?) in order to improve reactivity
- use of a javascript dependency management like bower would ease to add or update usefull js libs
- a new improved jukebox player currently under development #277

**Comment by muff1nman** [comment]
@biconou I think all those sound good with two caveats:

- I'm still not convinced we need bower right now
- We should be careful not to swap out too much js at once. Or if we do, perhaps we need the ability to have a switch between experimental vs stable ui.

**Comment by biconou** [comment]
@muff1nman 
- bower is not an absolute necessity but, despite the fact I have no experience with it, it seems to be simple to reach. What are you suspicious about ?
- Agree with your twice bullet point. Can we suggest to 

1. create different jsp fragments for each player technology
2. avoid mixing javascript with scriptlet/conditional jsp tags
3. move js code to a .js file

- Perhaps the 6.3 will be to short run to include the new jukebox. Improvements needs to be made and It really needs to be tested by someone else that me.

**Comment by muff1nman** [comment]
@biconou What about putting the switch at the controller level. I.e:

```
@RequestMapping
public String doGet() {
    ...
    if(experimental) {
        return "playQueue2"
    } else {
        return "playQueue"
    }
```

**Comment by muff1nman** [comment]
Not necessarily suspicious, just hesitant to take too large of steps and including bower seems like it would open the door for including a lot of modern frameworks all at once. I think as long as we take it easy, bower could be a good first step. Really I'm just reluctant to introduce a bunch of change at once unless we plan to do an entire rewrite.

**Comment by muff1nman** [comment]
Some suggestions for 6.3:

- Remove DWR
  - Its not really a best practice anymore
  - It doesn't make sense to use DWR and have a REST api
  - It will be needed if/when we rewrite the frontend
  - Would improve debugging experience as its kind of a black box.
- Unify authentication for rest apis and normal requets (will also address #258)
  - Blocks Removing DWR as the REST api will need to be used from the javascript via ajax calls
- Replace JWPlayer
  - MediaElement.js looks good

**Comment by biconou** [comment]
@muff1nman about bower, I can understand your arguments but on the other hand this project is well administered. 
I agree for DWR (even if I thought DWR was a nice solution years ago).
Another interesting js audio library is SoundManager (http://www.schillmania.com/projects/soundmanager2/).

**Comment by jooola** [comment]
Here are somes old issues, i think those needs to be done as well : 
* Upgrade Lucene
* Use ffprobe for metadata discovery

**Comment by biconou** [comment]
I think that upgrading Lucene is not useful without improving search functionalities.

**Comment by biconou** [comment]
@jooola What is the use of ffprobe ?

**Comment by biconou** [comment]
The new jukebox player will be ready for 6.3 (I work a lot on fixing it this week end). It will be still experimental as long as it does not replace the existing jukebox player.

**Comment by jooola** [comment]
Yeah sorry i didn't add links to the current issues : 
Upgrade Lucene #135 
Use ffprobe for metadata discovery #117

**Comment by biconou** [comment]
@muff1nman I would be glad to work on "Unify authentication for rest apis and normal requets"

**Comment by muff1nman** [comment]
I think I tested it and you should be able to invoke the rest API using the session cookie.

**Comment by biconou** [comment]
@muff1nman Yes, I tested it last night and it works. That's cool.

**Comment by muff1nman** [comment]
Yep, I think I'd like to start migrating some of the dwr stuff into the rest service

**Comment by biconou** [comment]
I could do it for the new jukebox. 
I made some tests last night and, as the the Libresonic API already support json responses, it's pretty easy.

Here is a simple JSP that pings the rest API and displays the API version in a div.
The session cookie is implicitly used for authentication.

```
<%@ page language="java" contentType="text/html; charset=utf-8" pageEncoding="iso-8859-1" %>

<html>
<head>
    <%@ include file="head.jsp" %>
    <%@ include file="jquery.jsp" %>
</head>

<script language"javascript">
$(document).ready(function() {
  $.get( "rest/ping.view?v=1.14.0&c=web&f=json", function( data ) {
    $(".result").html(data['subsonic-response'].version);
  });
});
</script>

<body class="mainframe bgcolor1" style="padding-bottom:0.5em">

<div class="result"></div>

</body>
</html>
```

**Comment by HeaDBanGer84** [comment]
bower is more or less deprecated. Try to move to [yarn](https://yarnpkg.com/lang/en/) works very similar but is more active in development.

**Comment by aairey** [comment]
Is this project abandoned (again)?
Back to airsonic as they have a more recent release?
https://github.com/airsonic/airsonic/releases
